modeling/local-slice-algorithm.go

```diff
@@ -0,0 +1,171 @@
+}
+
+// pop the front
+func (el *endpointsList) pop() {
+	if len(el.byZone) == 0 {
+		return
+	}
+	el.byZone = el.byZone[1:]
+}
+
+// CreateSliceGroups creates sliceGroups with 'one local EndpointSliceGroup per
+// zone' policy
+func (alg LocalSliceAlgorithm) CreateSliceGroups(region regionInfo) (map[string]EndpointSliceGroup, error) {
+	if region.zoneDetails == nil {
+		return nil, fmt.Errorf("zoneDetail should not be nil")
+	}
+	sliceGroups := map[string]EndpointSliceGroup{}
+	// endpointsAvailable stores 'endpointDeviation' -- zones with 'extra'
+	// endpoints
+	endpointsAvailable := endpointsList{}
+	// endpointsNeeded stores zones with number of endpoints needed
+	endpointsNeeded := endpointsList{}
+	for zoneName, zone := range region.zoneDetails {
+		var localGroup EndpointSliceGroup
+		localGroup.Label = zoneName
+		// this local sliceGroup should only receive traffic from current zone,
+		// this map tracks weights of traffic from different zones to this
+		// sliceGroup
+		localGroup.Weights = map[string]float64{zoneName: 1.0}
+		// this map stores composition of this sliceGroup, zoneName - number of
+		// endpoints from zoneName
+		localGroup.Composition = map[string]int{}
+
+		// the expected number of endpoints is with the same ratio as the number
+		// of nodes in this zone
+		expectedEndpoints := int(zone.nodesRatio * float64(region.totalEndpoints))
+		// deviation: - means need more endpoints from other sliceGroups, +
+		// means need give out endpoints to other sliceGroups
+		deviation := zone.Endpoints - expectedEndpoints
+		if deviation > 0 {
+			endpointsAvailable.push(endpointDeviation{name: zoneName, deviation: deviation})
+			localGroup.Composition[zoneName] = int(expectedEndpoints)
+		} else {
+			endpointsNeeded.push(endpointDeviation{name: zoneName, deviation: -deviation})
+			localGroup.Composition[zoneName] = zone.Endpoints
+		}
+
+		sliceGroups[zoneName] = localGroup
+	}
+
+	err := balanceSliceGroups(&endpointsAvailable, &endpointsNeeded, sliceGroups)
+	if err != nil {
+		return nil, err
+	}
+	return sliceGroups, nil
+}
+
+// balanceSliceGroups helps to 'rent' endpoints from zones have extra endpoints
+// to EndpointSliceGroups that have insufficient ones
+func balanceSliceGroups(endpointsAvailable *endpointsList, endpointsNeeded *endpointsList, sliceGroups map[string]EndpointSliceGroup) error {
+	for _, receiveZone := range endpointsNeeded.byZone {
+		// The available list is empty while there are still endpoints in
+		// need. This can happen when the approximation on deviation
+		// (calculated above) ends up in asymmetric sums of endpoints
+		// available and endpoints in need (in need > available)
+		if len(endpointsAvailable.byZone) == 0 {
+			makeupForApproximation(endpointsNeeded, sliceGroups, false)
+			return nil
+		}
+		// This is ugly, but the index below is a copy, after pop no way to
+		// udpate the index to its new position
+		indexDelta := 0
+		for index, sendZone := range endpointsAvailable.byZone {
```

Comment:
In this case it might be simpler to use a traditional for loop where you manually control/increment i instead of having to maintain this separate value.

Reply:
makes sense, change to traditional loop